### PR TITLE
#53 Testing the api endpoints

### PR DESCRIPTION
**Pull request by samyshyaka**
#### What does this PR do?
Test the endpoints in the REST API
#### Description of Task to be completed?
Have the following endpoints tested

GET api/v1/articles

GET api/v1/articles/:id

POST api/v1/articles

PUT api/v1/articles/:id

DELETE api/v1/articles/:id
#### How should this be manually tested?
Clone the repository

- cd into the repository
- git checkout development
- git fetch origin
- npm install
- npm run test
#### Any background context you want to provide?
Ensure you have a good internet for the api to be able to load the cloud database

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
   "scripts": {
     "start": "node src/app.js",
     "start:dev": "nodemon src/app.js",
-    "test": "mocha test",
+    "test": "mocha test --timeout 1000000000",
```
Comment (feyton) [inline package.json]: This number is astonishing. Please keep it under 30 seconds. It will workout fine. Worked on my end.

File: package.json [hunk 2]
```diff
@@ -10,7 +10,7 @@
   "scripts": {
     "start": "node src/app.js",
     "start:dev": "nodemon src/app.js",
```
Comment (feyton) [inline package.json:12]: This script. I don't understand how to start it. 
Is it so different from just using **"dev"**

File: src/config/dbConn.js [hunk 3]
```diff
@@ -2,6 +2,12 @@ import mongoose from 'mongoose';
 
 const connectDB = async () => {
     try {
+        if (process.env.NODE_ENV === 'test'){
+            await mongoose.connect("mongodb+srv://samuel:MiMM4TXX4kiofqhO@cluster0.uj3am.mongodb.net/test?retryWrites=true&w=majority", {
```
Comment (feyton) [inline src/config/dbConn.js]: These are among the **sensitive information**. You can add these to .env
Ensure that people will be requested to use their testing database.

File: test/articles.spec.js [hunk 4]
```diff
@@ -39,30 +40,32 @@ describe('/POST article', () => {
...
-    //             res.body.should.have.property('data');
-    //             res.body.data.should.have.property('article');
-    //             res.body.data.article.should.have.property('title');
-    //             res.body.data.article.should.have.property('author');
-    //             res.body.data.article.should.have.property('content');
-    //         done();
-    //       });
-    // });
+    it('it should POST an article ', (done) => {
+        let token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MWU2OWFlOTFmMDEyZjg2MDE4ODQwNTMiLCJuYW1lIjoiRXJpYyIsInBhc3N3b3JkIjoiJDJiJDEwJGJtSUtSUzN6UU56Y1pMaG0wM3NUZk9sVmFWcG16N2J5cnRydnNNMEV6WUguTlQuVXJqM25HIiwiX192IjowLCJpYXQiOjE2NDI1MTYzMjcsImV4cCI6MTY0MjUxOTkyN30.9QKmLYzEjrIxEndJ7HvAKMMwYZwFjgDeMupw2nTKBes"
```
Comment (feyton) [inline test/articles.spec.js]: This token never worked for me. How are you generating valid tokens. We can send a request using mocha and then use the token from the data being sent back.

File: test/articles.spec.js [hunk 5]
```diff
@@ -39,30 +40,32 @@ describe('/POST article', () => {
...
+                res.body.data.article.should.have.property('author');
+                res.body.data.article.should.have.property('content');
+            done();
+          });
+    });
 });
 describe('/GET/:id article', () => {
     it('it should GET the article by the given id', (done) => {
-        let id = "61e4014690b12add9ba7b332"
+        let id = "61e55e5379904e9bd63020af"
```
Comment (feyton) [inline test/articles.spec.js]: Same as for the token. We can get these from request responses. Check this snippet. Not sure if its good testing practice, but works so great.
```
  describe("User login before creating a post", async () => {
    it("It should return a token for the logged in user", async () => {
      let userInfo = {
        password: "Your password of choice",
        email: "A valid email or username",
      };
      const userResult = chai.request(app)

        .post(apiRoute + "/accounts/login")
        .send(userInfo);
      expect(userResult).to.have.status(200);
      token = userResult.body.data.token;
    });
  });
  describe("Creating a new post after login", async () => {
    it("Should return a newly created post", async () => {
      const postCreation = await chai
        .request(app)
        .post(apiRoute + "/blogs")
        .send(post)
        .set("Authorization", "Bearer " + token);
      expect(postCreation).to.have.status(200);
      expect(postCreation.body.data).to.have.property("slug");
      postId = postCreation.body.data._id;
    });
  });
```

**Review by feyton — CHANGES_REQUESTED** [review]
There are some ways you can improve on this code base. Especially in **testing** the end points.
This is great work. @samuel-shyaka-dus 
## Working with **async/await**
Something I learned the hard way. is to set the request response to a **_variable_** before calling await and then you add assertion on the variable.

### Example:
```
const userRegisterRequest = await chai.request(app).get("/)
expect(userregisterRequest.body).to.have.property("token")
```

File: package.json (hunk 1 above)
Comment (samyshyaka) [reply package.json]: Will do that

File: package.json (hunk 2 above)
Comment (samyshyaka) [reply package.json:12]: I think it is the same thing

File: src/config/dbConn.js (hunk 3 above)
Comment (samyshyaka) [reply src/config/dbConn.js]: Sure, will do that

File: test/articles.spec.js (hunk 4 above)
Comment (samyshyaka) [reply test/articles.spec.js]: I changed this. now, I'm sending a login request first and getting the token from the response which I then use as authorization to the post endpoint.

File: test/articles.spec.js (hunk 5 above)
Comment (samyshyaka) [reply test/articles.spec.js]: I changed this too. I am getting the id from the response. Will push my code and create another PR

File: test/articles.spec.js (hunk 4 above)
Comment (Jkadhuwa) [reply test/articles.spec.js]: When generating tokens to be used in other tests you can use the before hook:
let token;
 before(function() {
    token = chai.request(app).post(/routetoauthentication).send(logindata):
  });
and use the  token wherever you wish.

**Review by Jkadhuwa — COMMENTED** [review]
Nice work so far, work on the given feedback and it will be ready for merging.

**Review by feyton — APPROVED** [review]

File: test/articles.spec.js (hunk 4 above)
Comment (samyshyaka) [reply test/articles.spec.js]: Will try that

**Pull request closed without merging**